k-distribution/src/main/scripts/lib/pyk/kastManip.py

```diff
@@ -366,19 +363,19 @@ def _pushDownRewrites(_kast):
             rhs = _kast['rhs']
             if lhs == rhs:
                 return lhs
-            if isKVariable(lhs) and isKVariable(rhs) and lhs['name'] == rhs['name']:
+            if type(lhs) is KVariable and type(rhs) is KVariable and lhs.name == rhs.name:
                 return lhs
-            if isKApply(lhs) and isKApply(rhs) and lhs['label'] == rhs['label'] and len(lhs['args']) == len(rhs['args']):
-                newArgs = [KRewrite(lArg, rArg) for (lArg, rArg) in zip(lhs['args'], rhs['args'])]
-                return KApply(lhs['label'], newArgs)
+            if type(lhs) is KApply and type(rhs) is KApply and lhs.label == rhs.label and lhs.arity == rhs.arity:
+                newArgs = [KRewrite(lArg, rArg) for (lArg, rArg) in zip(lhs.args, rhs.args)]
+                return lhs.copy(args=newArgs)
```

Comment:
Ah, is it because sometimes we want to `copy` a `KApply` in this custom way?

Reply:
These classes are immutable (as far as Python makes such a thing possible), so copying their instances without any modification is not necessary. 

But perphaps `copy` is not the best name for this function after all. I chose this name as an analogous concept in Scala has that name. I'd call it `with`, but that's a reserved word in Python. @ehildenb, what do you think about `let`, e.g. `t2 = t1.let(token='T2')`?

Reply:
`let` sounds fine, or `.updatedWith(...)` is perhaps a bit more descriptive?